File: frontend/src/lib/schemas/paper-roll.ts
/**
 * Zod schemas for the Paper Rolls (bobinas de papel/filme) tier.
 *
 * Mirrors the Pydantic shapes from `backend/src/schemas/paper_roll.py`. Like
 * Fabric rolls, a `PaperRoll` carries an authoritative `current_meters` column
 * (NOT a ledger sum) plus a `paper_roll_movements` ledger for history. Decimal
 * columns serialize as strings on the wire, so metered fields are `z.string()`;
 * the form layer keeps them as strings and validates positivity via refinements.
 */

import { z } from "zod";

export const PAPER_TYPES = ["dtf_film", "sublimation_paper", "transfer_paper"] as const;
export const paperTypeSchema = z.enum(PAPER_TYPES);
export type PaperType = z.infer<typeof paperTypeSchema>;

export const PAPER_MOVEMENT_KINDS = ["entry", "exit", "adjustment"] as const;
export const paperMovementKindSchema = z.enum(PAPER_MOVEMENT_KINDS);
export type PaperMovementKind = z.infer<typeof paperMovementKindSchema>;

const pageMeta = {
  total: z.number(),
  page: z.number(),
  page_size: z.number(),
  has_more: z.boolean(),
};

// ---------- Catalog read ----------

export const paperRollReadSchema = z.object({
  id: z.string(),
  received_at: z.string(),
  supplier_name: z.string(),
  paper_type: paperTypeSchema,
  width_cm: z.number().int(),
  initial_meters: z.string(),
  current_meters: z.string(),
  consumed_meters: z.string(),
  min_stock: z.string().nullable().optional(),
  on_hand: z.string(),
  low_stock: z.boolean(),
  created_at: z.string(),
  updated_at: z.string(),
});
export type PaperRoll = z.infer<typeof paperRollReadSchema>;

export const paperRollPageSchema = z.object({
  items: z.array(paperRollReadSchema),
  ...pageMeta,
});
export type PaperRollPage = z.infer<typeof paperRollPageSchema>;

// ---------- Movements ledger ----------

export const paperRollMiniSchema = z.object({
  id: z.string(),
  paper_type: paperTypeSchema,
  supplier_name: z.string(),
});
export type PaperRollMini = z.infer<typeof paperRollMiniSchema>;

export const paperMovementReadSchema = z.object({
  id: z.string(),
  paper_roll_id: z.string(),
  paper_roll: paperRollMiniSchema.nullable().optional(),
  kind: paperMovementKindSchema,
  quantity: z.string(),
  notes: z.string().nullable(),
  created_at: z.string(),
});
export type PaperMovementRead = z.infer<typeof paperMovementReadSchema>;

export const paperMovementPageSchema = z.object({
  items: z.array(paperMovementReadSchema),
  ...pageMeta,
});
export type PaperMovementPage = z.infer<typeof paperMovementPageSchema>;

// ---------- Create / update payloads ----------

export type PaperRollCreate = {
  received_at: string;
  supplier_name: string;
  paper_type: PaperType;
  width_cm: number;
  initial_meters: string;
  current_meters?: string | null;
  min_stock?: string | null;
};
export type PaperRollUpdate = Partial<PaperRollCreate>;

export type PaperRollConsume = {
  quantity: string;
  notes?: string | null;
};

export type PaperMovementCreate = {
  paper_roll_id: string;
  kind: PaperMovementKind;
  quantity: string;
  notes?: string | null;
};

// ---------- Filters ----------

export type PaperRollFilters = {
  q?: string;
  paper_type?: PaperType;
  low_stock_only?: boolean;
  page?: number;
  page_size?: number;
};

export type PaperMovementFilters = {
  paper_roll_id?: string;
  kind?: PaperMovementKind;
  date_from?: string;
  date_to?: string;
  page?: number;
  page_size?: number;
};

// ---------- Form schemas ----------

const decimalString = z
  .string()
  .trim()
  .min(1, { message: "validation.required" })
  .refine((value) => /^\d+(\.\d+)?$/.test(value) || /^\d+,\d+$/.test(value), {
    message: "validation.numeric",
  })
  .transform((value) => value.replace(",", "."));

export const paperRollFormSchema = z.object({
  paper_type: paperTypeSchema,
  width_cm: z
    .string()
    .trim()
    .min(1, { message: "validation.required" })
    .refine((v) => /^\d+$/.test(v) && Number(v) > 0, { message: "validation.numeric" }),
  initial_meters: decimalString.refine((v) => Number(v) > 0, {
    message: "validation.positive",
  }),
  supplier_name: z.string().trim().min(1, { message: "validation.supplierRequired" }).max(120),
  received_at: z.string().trim().min(1, { message: "validation.required" }),
});
export type PaperRollFormValues = z.input<typeof paperRollFormSchema>;
export type PaperRollFormPayload = z.output<typeof paperRollFormSchema>;

File: backend/src/schemas/paper_roll.py
"""Pydantic schemas for the Paper Rolls (bobinas de papel/filme) feature.

Mirrors the Fabric (bobinas) feature: a ``PaperRoll`` carries an authoritative
``current_meters`` column (NOT a ledger sum) plus a ``paper_roll_movements``
ledger for traceable history. ``PaperRollRead`` adds a computed ``consumed_meters``
field and an ``on_hand`` alias (= ``current_meters``). Decimal columns are
serialized as strings (FastAPI / Pydantic default) — the frontend zod layer
mirrors that contract.

``low_stock`` is computed per row: a row-level ``min_stock`` (absolute meters
floor) overrides the company-wide ``paper`` threshold (default % of initial).
"""

from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from models.enums import PaperMovementKind, PaperType
from schemas._common import Page

# ---------- Catalog CRUD ----------


class PaperRollCreate(BaseModel):
    received_at: date
    supplier_name: str = Field(min_length=1, max_length=120)
    paper_type: PaperType
    width_cm: int = Field(gt=0)
    initial_meters: Decimal = Field(gt=Decimal("0"), max_digits=10, decimal_places=2)
    min_stock: Decimal | None = Field(default=None, ge=Decimal("0"), max_digits=10, decimal_places=2)
    # Optional — defaults to `initial_meters` at insert when omitted.
    current_meters: Decimal | None = Field(default=None, ge=Decimal("0"), max_digits=10, decimal_places=2)


class PaperRollUpdate(BaseModel):
    received_at: date | None = None
    supplier_name: str | None = Field(default=None, min_length=1, max_length=120)
    paper_type: PaperType | None = None
    width_cm: int | None = Field(default=None, gt=0)
    initial_meters: Decimal | None = Field(default=None, gt=Decimal("0"), max_digits=10, decimal_places=2)
    current_meters: Decimal | None = Field(default=None, ge=Decimal("0"), max_digits=10, decimal_places=2)
    min_stock: Decimal | None = Field(default=None, ge=Decimal("0"), max_digits=10, decimal_places=2)


class PaperRollRead(BaseModel):
    id: uuid.UUID
    received_at: date
    supplier_name: str
    paper_type: PaperType
    width_cm: int
    initial_meters: Decimal
    current_meters: Decimal
    consumed_meters: Decimal
    min_stock: Decimal | None = None
    on_hand: Decimal
    low_stock: bool
    created_at: datetime
    updated_at: datetime


# ---------- Roll-level consume ----------


class PaperRollConsume(BaseModel):
    quantity: Decimal = Field(gt=Decimal("0"), max_digits=12, decimal_places=3)
    notes: str | None = Field(default=None, max_length=500)


# ---------- Ledger movements ----------


class PaperMovementCreate(BaseModel):
    paper_roll_id: uuid.UUID
    kind: PaperMovementKind = Field(default=PaperMovementKind.ENTRY)
    quantity: Decimal = Field(gt=Decimal("0"), max_digits=12, decimal_places=3)
    notes: str | None = Field(default=None, max_length=500)


class PaperRollMini(BaseModel):
    id: uuid.UUID
    paper_type: PaperType
    supplier_name: str


class PaperMovementRead(BaseModel):
    id: uuid.UUID
    paper_roll_id: uuid.UUID
    paper_roll: PaperRollMini | None = None
    kind: PaperMovementKind
    quantity: Decimal
    notes: str | None
    created_at: datetime


# ---------- Filters ----------


class PaperRollFilters(BaseModel):
    q: str | None = Field(default=None, max_length=120)
    paper_type: PaperType | None = None
    low_stock_only: bool = False


class PaperMovementFilters(BaseModel):
    paper_roll_id: uuid.UUID | None = None
    kind: PaperMovementKind | None = None
    date_from: date | None = None
    date_to: date | None = None


# ---------- Page aliases ----------


PaperRollPage = Page[PaperRollRead]
PaperMovementPage = Page[PaperMovementRead]
